File: frontend/src/services/medicalRecordService.js
import api from './api';
import { getStoredRecords, saveStoredRecords } from './mockData';

export const medicalRecordService = {
  getAll: async (params = {}) => {
    try {
      const response = await api.get('/medical-records', { params });
      if (response.data && response.data.items) return response.data;
    } catch (e) { /* fallback below */ }

    let list = [...getStoredRecords()];
    const { search, gender, bloodGroup, department, status, sortBy = 'createdAt', sortOrder = 'desc', page = 1, pageSize = 10 } = params;

    if (search) {
      const s = search.toLowerCase();
      list = list.filter(r =>
        r.patientName.toLowerCase().includes(s) ||
        r.patientId.toLowerCase().includes(s) ||
        r.diagnosis.toLowerCase().includes(s) ||
        r.doctorName.toLowerCase().includes(s)
      );
    }
    if (gender) list = list.filter(r => r.gender.toLowerCase() === gender.toLowerCase());
    if (bloodGroup) list = list.filter(r => r.bloodGroup.toLowerCase() === bloodGroup.toLowerCase());
    if (department) list = list.filter(r => r.department.toLowerCase() === department.toLowerCase());
    if (status) list = list.filter(r => r.status.toLowerCase() === status.toLowerCase());

    list.sort((a, b) => {
      const valA = a[sortBy] ?? '';
      const valB = b[sortBy] ?? '';
      if (sortOrder === 'asc') return valA > valB ? 1 : -1;
      return valA < valB ? 1 : -1;
    });

    const totalCount = list.length;
    const items = list.slice((page - 1) * pageSize, page * pageSize);
    return {
      items,
      totalCount,
      page: Number(page),
      pageSize: Number(pageSize),
      totalPages: Math.ceil(totalCount / pageSize)
    };
  },

  getById: async (id) => {
    try {
      const response = await api.get(`/medical-records/${id}`);
      if (response.data && response.data.id) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    return list.find(r => r.id === Number(id)) || null;
  },

  getStats: async () => {
    try {
      const response = await api.get('/medical-records/stats');
      if (response.data && response.data.totalRecords !== undefined) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const depts = new Set(list.map(r => r.department));
    return {
      totalRecords: list.length,
      activePatients: list.filter(r => r.status === 'Active').length,
      completedCases: list.filter(r => r.status === 'Completed').length,
      dischargedPatients: list.filter(r => r.status === 'Discharged').length,
      pendingCases: list.filter(r => r.status === 'Pending').length,
      totalDepartments: depts.size
    };
  },

  create: async (data) => {
    try {
      const response = await api.post('/medical-records', data);
      if (response.data && response.data.id) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const newId = list.length > 0 ? Math.max(...list.map(r => r.id)) + 1 : 1;
    const record = {
      ...data,
      id: newId,
      createdBy: 'subadmin',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    };
    list.unshift(record);
    saveStoredRecords(list);
    return record;
  },

  update: async (id, data) => {
    try {
      const response = await api.put(`/medical-records/${id}`, data);
      if (response.data && response.data.id) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const idx = list.findIndex(r => r.id === Number(id));
    if (idx !== -1) {
      list[idx] = { ...list[idx], ...data, updatedAt: new Date().toISOString() };
      saveStoredRecords(list);
      return list[idx];
    }
    return null;
  },

  delete: async (id) => {
    try {
      const response = await api.delete(`/medical-records/${id}`);
      if (response.status === 200) return true;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const filtered = list.filter(r => r.id !== Number(id));
    saveStoredRecords(filtered);
    return true;
  },

  getPatients: async (params = {}) => {
    try {
      const response = await api.get('/medical-records/patients', { params });
      if (response.data && response.data.items) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const groups = {};
    for (const r of list) {
      if (!groups[r.patientId]) {
        groups[r.patientId] = [];
      }
      groups[r.patientId].push(r);
    }

    let patientSummaries = Object.keys(groups).map(pid => {
      const recs = groups[pid];
      recs.sort((a, b) => new Date(b.admissionDate) - new Date(a.admissionDate));
      const latest = recs[0];
      return {
        patientId: pid,
        patientName: latest.patientName,
        age: latest.age,
        gender: latest.gender,
        bloodGroup: latest.bloodGroup,
        recordCount: recs.length,
        latestAdmissionDate: latest.admissionDate,
        latestStatus: latest.status,
        latestDiagnosis: latest.diagnosis,
        latestDoctor: latest.doctorName,
        department: latest.department
      };
    });

    const { search, page = 1, pageSize = 8 } = params;
    if (search) {
      const s = search.toLowerCase();
      patientSummaries = patientSummaries.filter(p =>
        p.patientName.toLowerCase().includes(s) ||
        p.patientId.toLowerCase().includes(s) ||
        p.latestDiagnosis.toLowerCase().includes(s) ||
        p.latestDoctor.toLowerCase().includes(s)
      );
    }

    const totalCount = patientSummaries.length;
    const items = patientSummaries.slice((page - 1) * pageSize, page * pageSize);
    return {
      items,
      totalCount,
      page: Number(page),
      pageSize: Number(pageSize),
      totalPages: Math.ceil(totalCount / pageSize)
    };
  },

  getPatientHistory: async (patientId) => {
    try {
      const response = await api.get(`/medical-records/patients/${patientId}`);
      if (response.data && response.data.records) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const records = list
      .filter(r => r.patientId.toLowerCase() === patientId.toLowerCase())
      .sort((a, b) => new Date(b.admissionDate) - new Date(a.admissionDate));

    if (records.length === 0) return null;
    const latest = records[0];
    return {
      summary: {
        patientId: latest.patientId,
        patientName: latest.patientName,
        age: latest.age,
        gender: latest.gender,
        bloodGroup: latest.bloodGroup,
        recordCount: records.length,
        latestAdmissionDate: latest.admissionDate,
        latestStatus: latest.status,
        latestDiagnosis: latest.diagnosis,
        latestDoctor: latest.doctorName,
        department: latest.department
      },
      records
    };
  },

  getAnalytics: async () => {
    try {
      const response = await api.get('/medical-records/analytics');
      if (response.data && response.data.totalRecords !== undefined) return response.data;
    } catch (e) { /* fallback */ }

    const list = getStoredRecords();
    const total = list.length;
    const patients = new Set(list.map(r => r.patientId));

    const calcDistribution = (field) => {
      const counts = {};
      list.forEach(r => { counts[r[field]] = (counts[r[field]] || 0) + 1; });
      return Object.keys(counts)
        .map(name => ({
          name,
          count: counts[name],
          percentage: total > 0 ? Math.round((counts[name] / total) * 1000) / 10 : 0
        }))
        .sort((a, b) => b.count - a.count);
    };

    const ageBrackets = [
      { name: "0-18 (Pediatric)", count: list.filter(r => r.age <= 18).length },
      { name: "19-35 (Young Adult)", count: list.filter(r => r.age >= 19 && r.age <= 35).length },
      { name: "36-50 (Middle Age)", count: list.filter(r => r.age >= 36 && r.age <= 50).length },
      { name: "51-65 (Mature Adult)", count: list.filter(r => r.age >= 51 && r.age <= 65).length },
      { name: "65+ (Senior)", count: list.filter(r => r.age > 65).length }
    ];
    ageBrackets.forEach(a => {
      a.percentage = total > 0 ? Math.round((a.count / total) * 1000) / 10 : 0;
    });

    return {
      totalRecords: total,
      totalPatients: patients.size,
      activeCases: list.filter(r => r.status === 'Active').length,
      completedCases: list.filter(r => r.status === 'Completed').length,
      dischargedCases: list.filter(r => r.status === 'Discharged').length,
      pendingCases: list.filter(r => r.status === 'Pending').length,
      departmentDistribution: calcDistribution('department'),
      statusDistribution: calcDistribution('status'),
      genderDistribution: calcDistribution('gender'),
      bloodGroupDistribution: calcDistribution('bloodGroup'),
      ageGroupDistribution: ageBrackets
    };
  },

  getExportData: async (params = {}) => {
    try {
      const response = await api.get('/medical-records/export', { params });
      if (response.data && Array.isArray(response.data)) return response.data;
    } catch (e) { /* fallback */ }

    let list = [...getStoredRecords()];
    const { search, department, status } = params;
    if (search) {
      const s = search.toLowerCase();
      list = list.filter(r =>
        r.patientName.toLowerCase().includes(s) ||
        r.patientId.toLowerCase().includes(s) ||
        r.diagnosis.toLowerCase().includes(s) ||
        r.doctorName.toLowerCase().includes(s)
      );
    }
    if (department) list = list.filter(r => r.department.toLowerCase() === department.toLowerCase());
    if (status) list = list.filter(r => r.status.toLowerCase() === status.toLowerCase());
    return list;
  },
};

File: backend/Services/MedicalRecordService.cs
using Microsoft.EntityFrameworkCore;
using MedicalAdminPortal.Data;
using MedicalAdminPortal.DTOs;
using MedicalAdminPortal.Models;

namespace MedicalAdminPortal.Services;

public interface IMedicalRecordService
{
    Task<PaginatedResult<MedicalRecordDto>> GetAllAsync(
        string? search, string? gender, string? bloodGroup,
        string? department, string? status, string? sortBy,
        string? sortOrder, int page, int pageSize);
    Task<MedicalRecordDto?> GetByIdAsync(int id);
    Task<MedicalRecordDto> CreateAsync(CreateMedicalRecordRequest request, string createdBy);
    Task<MedicalRecordDto?> UpdateAsync(int id, UpdateMedicalRecordRequest request);
    Task<bool> DeleteAsync(int id);
    Task<DashboardStatsDto> GetStatsAsync();
    Task<PaginatedResult<PatientSummaryDto>> GetPatientsAsync(string? search, int page, int pageSize);
    Task<PatientHistoryDto?> GetPatientHistoryAsync(string patientId);
    Task<AnalyticsDto> GetDetailedAnalyticsAsync();
    Task<List<MedicalRecordDto>> GetAllForExportAsync(string? search, string? department, string? status);
}

public class MedicalRecordService : IMedicalRecordService
{
    private readonly ApplicationDbContext _context;

    public MedicalRecordService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<PaginatedResult<MedicalRecordDto>> GetAllAsync(
        string? search, string? gender, string? bloodGroup,
        string? department, string? status, string? sortBy,
        string? sortOrder, int page, int pageSize)
    {
        var query = _context.MedicalRecords.AsQueryable();

        // Search
        if (!string.IsNullOrWhiteSpace(search))
        {
            var searchLower = search.ToLower();
            query = query.Where(r =>
                r.PatientName.ToLower().Contains(searchLower) ||
                r.PatientId.ToLower().Contains(searchLower) ||
                r.Diagnosis.ToLower().Contains(searchLower) ||
                r.DoctorName.ToLower().Contains(searchLower));
        }

        // Filters
        if (!string.IsNullOrWhiteSpace(gender) && Enum.TryParse<Gender>(gender, true, out var genderEnum))
        {
            query = query.Where(r => r.Gender == genderEnum);
        }

        if (!string.IsNullOrWhiteSpace(bloodGroup) && Enum.TryParse<BloodGroup>(bloodGroup, true, out var bgEnum))
        {
            query = query.Where(r => r.BloodGroup == bgEnum);
        }

        if (!string.IsNullOrWhiteSpace(department) && Enum.TryParse<Department>(department, true, out var deptEnum))
        {
            query = query.Where(r => r.Department == deptEnum);
        }

        if (!string.IsNullOrWhiteSpace(status) && Enum.TryParse<RecordStatus>(status, true, out var statusEnum))
        {
            query = query.Where(r => r.Status == statusEnum);
        }

        // Total count before pagination
        var totalCount = await query.CountAsync();

        // Sorting
        query = (sortBy?.ToLower(), sortOrder?.ToLower()) switch
        {
            ("patientname", "asc") => query.OrderBy(r => r.PatientName),
            ("patientname", _) => query.OrderByDescending(r => r.PatientName),
            ("age", "asc") => query.OrderBy(r => r.Age),
            ("age", _) => query.OrderByDescending(r => r.Age),
            ("admissiondate", "asc") => query.OrderBy(r => r.AdmissionDate),
            ("admissiondate", _) => query.OrderByDescending(r => r.AdmissionDate),
            ("department", "asc") => query.OrderBy(r => r.Department),
            ("department", _) => query.OrderByDescending(r => r.Department),
            ("createdat", "asc") => query.OrderBy(r => r.CreatedAt),
            ("createdat", _) => query.OrderByDescending(r => r.CreatedAt),
            ("status", "asc") => query.OrderBy(r => r.Status),
            ("status", _) => query.OrderByDescending(r => r.Status),
            _ => query.OrderByDescending(r => r.CreatedAt)
        };

        // Pagination
        var items = await query
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(r => MapToDto(r))
            .ToListAsync();

        return new PaginatedResult<MedicalRecordDto>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<MedicalRecordDto?> GetByIdAsync(int id)
    {
        var record = await _context.MedicalRecords.FindAsync(id);
        return record == null ? null : MapToDto(record);
    }

    public async Task<MedicalRecordDto> CreateAsync(CreateMedicalRecordRequest request, string createdBy)
    {
        var record = new MedicalRecord
        {
            PatientId = request.PatientId,
            PatientName = request.PatientName,
            Age = request.Age,
            Gender = Enum.Parse<Gender>(request.Gender, true),
            BloodGroup = Enum.Parse<BloodGroup>(request.BloodGroup, true),
            Diagnosis = request.Diagnosis,
            DoctorName = request.DoctorName,
            Department = Enum.Parse<Department>(request.Department, true),
            AdmissionDate = request.AdmissionDate,
            DischargeDate = request.DischargeDate,
            Status = Enum.Parse<RecordStatus>(request.Status, true),
            Notes = request.Notes,
            CreatedBy = createdBy,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _context.MedicalRecords.Add(record);
        await _context.SaveChangesAsync();

        return MapToDto(record);
    }

    public async Task<MedicalRecordDto?> UpdateAsync(int id, UpdateMedicalRecordRequest request)
    {
        var record = await _context.MedicalRecords.FindAsync(id);
        if (record == null) return null;

        record.PatientId = request.PatientId;
        record.PatientName = request.PatientName;
        record.Age = request.Age;
        record.Gender = Enum.Parse<Gender>(request.Gender, true);
        record.BloodGroup = Enum.Parse<BloodGroup>(request.BloodGroup, true);
        record.Diagnosis = request.Diagnosis;
        record.DoctorName = request.DoctorName;
        record.Department = Enum.Parse<Department>(request.Department, true);
        record.AdmissionDate = request.AdmissionDate;
        record.DischargeDate = request.DischargeDate;
        record.Status = Enum.Parse<RecordStatus>(request.Status, true);
        record.Notes = request.Notes;
        record.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();

        return MapToDto(record);
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var record = await _context.MedicalRecords.FindAsync(id);
        if (record == null) return false;

        _context.MedicalRecords.Remove(record);
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<DashboardStatsDto> GetStatsAsync()
    {
        var records = _context.MedicalRecords;

        return new DashboardStatsDto
        {
            TotalRecords = await records.CountAsync(),
            ActivePatients = await records.CountAsync(r => r.Status == RecordStatus.Active),
            CompletedCases = await records.CountAsync(r => r.Status == RecordStatus.Completed),
            DischargedPatients = await records.CountAsync(r => r.Status == RecordStatus.Discharged),
            PendingCases = await records.CountAsync(r => r.Status == RecordStatus.Pending),
            TotalDepartments = await records.Select(r => r.Department).Distinct().CountAsync()
        };
    }

    public async Task<PaginatedResult<PatientSummaryDto>> GetPatientsAsync(string? search, int page, int pageSize)
    {
        var records = await _context.MedicalRecords.ToListAsync();

        var grouped = records
            .GroupBy(r => r.PatientId)
            .Select(g =>
            {
                var latest = g.OrderByDescending(r => r.AdmissionDate).First();
                return new PatientSummaryDto
                {
                    PatientId = g.Key,
                    PatientName = latest.PatientName,
                    Age = latest.Age,
                    Gender = latest.Gender.ToString(),
                    BloodGroup = FormatBloodGroup(latest.BloodGroup),
                    RecordCount = g.Count(),
                    LatestAdmissionDate = latest.AdmissionDate,
                    LatestStatus = latest.Status.ToString(),
                    LatestDiagnosis = latest.Diagnosis,
                    LatestDoctor = latest.DoctorName,
                    Department = FormatDepartment(latest.Department)
                };
            })
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(search))
        {
            var s = search.ToLower();
            grouped = grouped.Where(p =>
                p.PatientName.ToLower().Contains(s) ||
                p.PatientId.ToLower().Contains(s) ||
                p.LatestDiagnosis.ToLower().Contains(s) ||
                p.LatestDoctor.ToLower().Contains(s));
        }

        var totalCount = grouped.Count();
        var items = grouped
            .OrderByDescending(p => p.LatestAdmissionDate)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToList();

        return new PaginatedResult<PatientSummaryDto>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<PatientHistoryDto?> GetPatientHistoryAsync(string patientId)
    {
        var records = await _context.MedicalRecords
            .Where(r => r.PatientId == patientId)
            .OrderByDescending(r => r.AdmissionDate)
            .ToListAsync();

        if (!records.Any()) return null;

        var latest = records.First();
        var summary = new PatientSummaryDto
        {
            PatientId = latest.PatientId,
            PatientName = latest.PatientName,
            Age = latest.Age,
            Gender = latest.Gender.ToString(),
            BloodGroup = FormatBloodGroup(latest.BloodGroup),
            RecordCount = records.Count,
            LatestAdmissionDate = latest.AdmissionDate,
            LatestStatus = latest.Status.ToString(),
            LatestDiagnosis = latest.Diagnosis,
            LatestDoctor = latest.DoctorName,
            Department = FormatDepartment(latest.Department)
        };

        return new PatientHistoryDto
        {
            Summary = summary,
            Records = records.Select(MapToDto).ToList()
        };
    }

    public async Task<AnalyticsDto> GetDetailedAnalyticsAsync()
    {
        var records = await _context.MedicalRecords.ToListAsync();
        var total = records.Count;
        var totalPatients = records.Select(r => r.PatientId).Distinct().Count();

        var deptGroups = records.GroupBy(r => r.Department)
            .Select(g => new NameCountDto
            {
                Name = FormatDepartment(g.Key),
                Count = g.Count(),
                Percentage = total > 0 ? Math.Round((double)g.Count() / total * 100, 1) : 0
            })
            .OrderByDescending(x => x.Count)
            .ToList();

        var statusGroups = records.GroupBy(r => r.Status)
            .Select(g => new NameCountDto
            {
                Name = g.Key.ToString(),
                Count = g.Count(),
                Percentage = total > 0 ? Math.Round((double)g.Count() / total * 100, 1) : 0
            })
            .OrderByDescending(x => x.Count)
            .ToList();

        var genderGroups = records.GroupBy(r => r.Gender)
            .Select(g => new NameCountDto
            {
                Name = g.Key.ToString(),
                Count = g.Count(),
                Percentage = total > 0 ? Math.Round((double)g.Count() / total * 100, 1) : 0
            })
            .OrderByDescending(x => x.Count)
            .ToList();

        var bloodGroups = records.GroupBy(r => r.BloodGroup)
            .Select(g => new NameCountDto
            {
                Name = FormatBloodGroup(g.Key),
                Count = g.Count(),
                Percentage = total > 0 ? Math.Round((double)g.Count() / total * 100, 1) : 0
            })
            .OrderByDescending(x => x.Count)
            .ToList();

        var ageGroups = new List<NameCountDto>
        {
            new() { Name = "0-18 (Pediatric)", Count = records.Count(r => r.Age <= 18) },
            new() { Name = "19-35 (Young Adult)", Count = records.Count(r => r.Age >= 19 && r.Age <= 35) },
            new() { Name = "36-50 (Middle Age)", Count = records.Count(r => r.Age >= 36 && r.Age <= 50) },
            new() { Name = "51-65 (Mature Adult)", Count = records.Count(r => r.Age >= 51 && r.Age <= 65) },
            new() { Name = "65+ (Senior)", Count = records.Count(r => r.Age > 65) }
        };
        foreach (var ag in ageGroups)
        {
            ag.Percentage = total > 0 ? Math.Round((double)ag.Count / total * 100, 1) : 0;
        }

        return new AnalyticsDto
        {
            TotalRecords = total,
            TotalPatients = totalPatients,
            ActiveCases = records.Count(r => r.Status == RecordStatus.Active),
            CompletedCases = records.Count(r => r.Status == RecordStatus.Completed),
            DischargedCases = records.Count(r => r.Status == RecordStatus.Discharged),
            PendingCases = records.Count(r => r.Status == RecordStatus.Pending),
            DepartmentDistribution = deptGroups,
            StatusDistribution = statusGroups,
            GenderDistribution = genderGroups,
            BloodGroupDistribution = bloodGroups,
            AgeGroupDistribution = ageGroups
        };
    }

    public async Task<List<MedicalRecordDto>> GetAllForExportAsync(string? search, string? department, string? status)
    {
        var query = _context.MedicalRecords.AsQueryable();

        if (!string.IsNullOrWhiteSpace(search))
        {
            var searchLower = search.ToLower();
            query = query.Where(r =>
                r.PatientName.ToLower().Contains(searchLower) ||
                r.PatientId.ToLower().Contains(searchLower) ||
                r.Diagnosis.ToLower().Contains(searchLower) ||
                r.DoctorName.ToLower().Contains(searchLower));
        }

        if (!string.IsNullOrWhiteSpace(department) && Enum.TryParse<Department>(department, true, out var deptEnum))
        {
            query = query.Where(r => r.Department == deptEnum);
        }

        if (!string.IsNullOrWhiteSpace(status) && Enum.TryParse<RecordStatus>(status, true, out var statusEnum))
        {
            query = query.Where(r => r.Status == statusEnum);
        }

        var list = await query.OrderByDescending(r => r.AdmissionDate).Take(1000).ToListAsync();
        return list.Select(MapToDto).ToList();
    }

    private static MedicalRecordDto MapToDto(MedicalRecord record)
    {
        return new MedicalRecordDto
        {
            Id = record.Id,
            PatientId = record.PatientId,
            PatientName = record.PatientName,
            Age = record.Age,
            Gender = record.Gender.ToString(),
            BloodGroup = FormatBloodGroup(record.BloodGroup),
            Diagnosis = record.Diagnosis,
            DoctorName = record.DoctorName,
            Department = FormatDepartment(record.Department),
            AdmissionDate = record.AdmissionDate,
            DischargeDate = record.DischargeDate,
            Status = record.Status.ToString(),
            Notes = record.Notes,
            CreatedBy = record.CreatedBy,
            CreatedAt = record.CreatedAt,
            UpdatedAt = record.UpdatedAt
        };
    }

    private static string FormatBloodGroup(BloodGroup bg)
    {
        return bg switch
        {
            BloodGroup.APositive => "A+",
            BloodGroup.ANegative => "A-",
            BloodGroup.BPositive => "B+",
            BloodGroup.BNegative => "B-",
            BloodGroup.ABPositive => "AB+",
            BloodGroup.ABNegative => "AB-",
            BloodGroup.OPositive => "O+",
            BloodGroup.ONegative => "O-",
            _ => bg.ToString()
        };
    }

    private static string FormatDepartment(Department dept)
    {
        return dept switch
        {
            Department.GeneralMedicine => "General Medicine",
            Department.EmergencyMedicine => "Emergency Medicine",
            _ => System.Text.RegularExpressions.Regex.Replace(dept.ToString(), "(\\B[A-Z])", " $1")
        };
    }
}
